### PR TITLE
Feature/remodel participations

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] ~I added integration test cases for the client (Jest) related to the features~ There is nothing new in the client, the existing tests were updated.
- [x] ~I added screenshots/screencast of my UI changes~
- [x] ~I translated all the newly inserted strings~

### Motivation and Context
After adding templateParticipation and solutionParticipation to ProgrammingExercise, there have been errors due to wrong usage of the Participation class. Therefore it was necessary to split the different Participations into a taxonomy using inheritance. The big advantage is that we have much more type safety in the future without the need of a large database refactoring ;-) 
The small disadvantage is that we need some additional casts.

### Description
We now have an abstract class Participation. All exercises except ProgrammingExercise can use the StudentParticipation from now on and don't need to deal with buildPlanId and repositoryUrl any more.
For Programming Exercises, we now have a ProgrammingExerciseStudentParticipation that adds buildPlanId and repositoryUrl to StudentParticipation and TemplateProgrammingExerciseParticipation as well as SolutionProgrammingExerciseParticipation which inherit directly from the abstract class Participation so that they are not connected to a student any more.

In addition this also prevents that template and solution participation are accidentally added to Exercise.participations (which happened before and led to some weird problems)

The good news is that the database changes are minimal due to the SingleTable strategy. We simply add a new column discriminator and set the correct values for all entries. While doing that we also remove wrong connections for template and solution participation to exercise.

### Steps for Testing
Basically all functionality in Artemis needs to be tested because Participation is a central concept that is used in many different places.

### CODE REVIEW

**Comment by thilo-behnke** [comment]
Giving this a first rough look, this looks very good! I think this is a good approach to make the modeling of participations more clear.

One thing that immediately caused issues for me: When querying the `programmingExercise` together with the template and solution participation, the endpoint fails because it can't serialize the circular dependency (programmingExercise -> templateParticipation -> programmingExercise), even though the `JsonIgnoreProperties` setting is correctly set.
This is e.g. the case in `/api/courses/<course-number>/programming-exercises/`.

I'll do a more detailed review later this week.

**Comment by jpbernius** [comment]
I started looking into this briefly and I do think this makes for a better understanding of a participation.

There might be some more possibilities to make relations to e.g. `StudentParticipation`s instead of `Participation`s on an domain class level, so we can remove a few more Type Casts?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 1]
```diff
@@ -427,15 +430,22 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
...
+            Set<StudentParticipation> participations = new HashSet<>();
             assignmentParticipation.ifPresent(participations::add);
             programmingExercise.setParticipations(participations);
 
             if (!authCheckService.isAtLeastInstructorForCourse(course, user)) {
                 return forbidden();
             }
 
+            // Avoid circular serialization issues with jackson.
+            if (programmingExercise.getTemplateParticipation() != null) {
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]: I committed this as a hotfix. If this is not done, there will be a circular serialization issue, even though `@JsonIgnoreProperties` is set on the programmingExercise in the Template/SolutionParticipation classes.

Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]: I committed this as a hotfix. If this is not done, there will be a circular serialization issue, even though @JsonIgnoreProperties is set on the programmingExercise in the Template/SolutionParticipation classes.

File: src/main/java/de/tum/in/www1/artemis/web/rest/QuizSubmissionResource.java [hunk 2]
```diff
@@ -153,7 +153,7 @@ public QuizSubmissionResource(QuizExerciseService quizExerciseService, QuizSubmi
         quizSubmission.calculateAndUpdateScores(quizExercise);
 
         // create Participation stub
-        Participation participation = new Participation().exercise(quizExercise);
+        StudentParticipation participation = new StudentParticipation().exercise(quizExercise);
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/QuizSubmissionResource.java]: I think this is an example that shows that we can be specific in some places about the Participations: This can never a Template/SolutionParticipation, so it's better to be clear here :+1: .

File: src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java [hunk 3]
```diff
@@ -409,7 +436,7 @@ else if (planKey.toLowerCase().contains("-solution")) { // TODO: transfer this i
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<Result> getLatestResultWithFeedbacks(@PathVariable Long participationId) {
         log.debug("REST request to get latest result for participation : {}", participationId);
-        Participation participation = participationService.findOne(participationId);
+        StudentParticipation participation = participationService.findOne(participationId);
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java]: All endpoints that now explicitly instantiate the StudentParticipation don't work anymore for Template/SolutionParticipations.
Besides new endpoints his will necessitate work in the programming exercise UIs (edit, view, edit in editor) because we have to make explicit requests to the different endpoints.

File: src/main/webapp/app/entities/participation/participation-websocket.service.ts [hunk 4]
```diff
@@ -20,7 +21,7 @@ export interface IParticipationWebsocketService {
 
 @Injectable({ providedIn: 'root' })
 export class ParticipationWebsocketService implements IParticipationWebsocketService {
-    cachedParticipations: Map<number /* ID of participation */, Participation> = new Map<number, Participation>();
+    cachedParticipations: Map<number /* ID of participation */, StudentParticipation> = new Map<number, StudentParticipation>();
```
Comment (thilo-behnke) [inline src/main/webapp/app/entities/participation/participation-websocket.service.ts]: This also means we need a new websocket service for the Template/SolutionParticipations.
So this would mean we need a generic ParticipationWebsocketService with a type parameter for the Participation to make reuse possible. This also means that we need different websocket channels for transporting participation messages for the 3 types.

File: src/main/java/de/tum/in/www1/artemis/domain/SolutionProgrammingExerciseParticipation.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import javax.persistence.*;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonView;
+import de.tum.in.www1.artemis.domain.view.QuizView;
+
+@Entity
+@DiscriminatorValue(value = "SPEP")
+public class SolutionProgrammingExerciseParticipation extends Participation implements ProgrammingExerciseParticipation {
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/domain/SolutionProgrammingExerciseParticipation.java]: I think it's good to also differentiate between the Template/Solution Participation in some instances, e.g. when creating the test cases from a Solution Participation's build result. :+1:

File: src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java (hunk 3 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java]: As discussed: It would save us a lot of time if the Result Endpoints would be Participation neutral.
Otherwise a lot of neutral GUI components in our client will have to differentiate the different participation types, e.g. when retrieving the latest result for the updating result component in the code-editor.

File: src/main/webapp/app/entities/participation/participation-websocket.service.ts (hunk 4 above)
Comment (thilo-behnke) [reply src/main/webapp/app/entities/participation/participation-websocket.service.ts]: Thinking about this again, my comment was incorrect: we don't use the participation caching for template / solution participations, just for receiving new results. So it can stay as is.

**Comment by thilo-behnke** [comment]
FYI: I'll be working on the fixes for the programming exercises the next days (I plan to finish it until friday).
This could cause the build to fail from now and then.

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryTemplateParticipationResource.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    public RepositoryTemplateParticipationResource(UserService userService, AuthorizationCheckService authCheckService, Optional<GitService> gitService,
+            Optional<ContinuousIntegrationService> continuousIntegrationService, RepositoryService repositoryService,
+            ProgrammingExerciseParticipationService programmingExerciseParticipationService) {
+        super(userService, authCheckService, gitService, continuousIntegrationService, repositoryService);
+        this.participationService = programmingExerciseParticipationService;
+    }
+
+    @Override
+    Repository getRepository(Long participationId) throws IOException, IllegalAccessException, InterruptedException {
+        Optional<TemplateProgrammingExerciseParticipation> participation = participationService.findTemplateParticipation(participationId);
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryTemplateParticipationResource.java]: This is an example where we now have duplication:
Because we can't get all programming exercise participation types from the same repository, we now have to use different endpoints for differentiation.

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 7]
```diff
@@ -133,18 +131,22 @@ public void onResultNotifiedNew(Participation participation, Object requestBody)
...
         }
     }
 
-    private void notifyUser(Participation participation, Result result) {
+    @Transactional(readOnly = true)
+    public void notifyUser(ProgrammingExerciseParticipation participation, Result result) {
         if (result != null) {
+            // Avoid circular serialization issues.
+            // TODO: This results in removing the reference from the participation to the exercise.
+            // result.getParticipation().setExercise(null);
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java]: Here is the issue that `@JsonIgnoreProperties` does not work for the Discriminator classes.
Therefore we have to set the exercise to null.

However as this method is called in a Transactional Context, the reference will be overriden.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 8]
```diff
@@ -28,37 +31,50 @@
...
 
     public void notifyPush(Long participationId, Object requestBody) {
-        Optional<Participation> optionalParticipation = participationRepository.findById(participationId);
-        if (!optionalParticipation.isPresent()) {
-            log.warn("Invalid participation received while notifying about push: " + participationId);
-            return;
+        ProgrammingExerciseParticipation participation;
+        // We don't know here what kind of participation to expect, so the only way is to try each one.
+        // TODO: This seems like a very bad way to handle this, but is there another way without changing the bitbucket / bamboo configuration?
+        try {
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java]: Here we only get a participationId, so we don't know from which repository to retrieve the participation.
(The implementation with try catch is just for illustration)

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java (hunk 7 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ResultService.java]: We also have this issue in other Resources that are not in a Transactional Context, e.g. ParticipationResource -> initParticipation.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java (hunk 1 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]: This issue was resolved by setting `@JsonIgnore` on the exercise getter of `ProgrammingExerciseTemplateParticipation` and `ProgrammingExerciseSolutionParticipation`.

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryTemplateParticipationResource.java (hunk 6 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryTemplateParticipationResource.java]: This can be resolved by implementing a generic ParticipationRepository and type casting when needed. Thanks @krusche

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java (hunk 7 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ResultService.java]: See above note about `@JsonIgnore`.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java (hunk 8 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java]: This can be resolved with the generic ParticipationRepository.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 9]
```diff
@@ -28,40 +27,46 @@
...
-        Participation participation = optionalParticipation.get();
-        if (participation.getInitializationState() == InitializationState.INACTIVE) {
+    public void notifyPush(Long participationId, Object requestBody) throws IllegalArgumentException {
+        Participation participation = participationService.findOne(participationId);
+        if (!(participation instanceof ProgrammingExerciseParticipation))
+            throw new IllegalArgumentException();
+
+        ProgrammingExerciseParticipation programmingExerciseParticipation = (ProgrammingExerciseParticipation) participation;
+
+        if (participation instanceof ProgrammingExerciseStudentParticipation
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java]: It should be impossible in my opinion that a template or solution build plan is deleted. This is why `resumeExercise` and `triggerBuild` will not be called here for template or solution participations.

Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java]: There would be errors anyway when going through `resumeExercise`, as the logic will try to copy the template participation's build plan.

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 10]
```diff
@@ -210,7 +210,9 @@ public void createNewManualResult(Result result, boolean isProgrammingExerciseWi
                 Hibernate.initialize(savedResult.getParticipation().getExercise());
             }
 
-            ltiService.onNewBuildResult((ProgrammingExerciseStudentParticipation) savedResult.getParticipation());
+            if (savedResult.getParticipation() instanceof ProgrammingExerciseStudentParticipation) {
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java]: I think here we also have to watch out that template or solution participations don't enter this area.

**Comment by thilo-behnke** [comment]
From my point of view, this should be done.
However I need to test this on our server first, but this PR could break our server until it is merged as the migration will affect newly created Participations.

**Comment by thilo-behnke** [comment]
@krusche I tested the programming exercises on the server and fixed the last issue I found.

**Comment by krusche** [comment]
What is the status of this PR? Could we merge it or should we test it more thoroughly and potentially also write some additional tests that can be executed automatically?

**Comment by thilo-behnke** [comment]
> What is the status of this PR? Could we merge it or should we test it more thoroughly and potentially also write some additional tests that can be executed automatically?

From my side this is done and tested, but we should have someone else do a review of the code (I had to touch a lot of files to fix the issues for the programming exercises).
About the tests: There is no new logic introduced, so I would not write tests for this PR. I made sure that all the existing tests are updated with the participation modeling.

**Comment by thilo-behnke** [comment]
I changed my mind about the tests, I'll write some now.

**Comment by thilo-behnke** [comment]
@krusche I added test cases for the new programming exercise participation resource.
Maybe we can have a look at this PR again on Tuesday together so we can merge it :)

**Comment by thilo-behnke** [comment]
The last merge with the checkout conflict resolution was pretty heavy, so I'll need to test this again on the server.

**Comment by thilo-behnke** [comment]
I tested this completely on the server again, I found the following issues:
Issues with Remodel Participation PR:
- [ ] Programming Exercises can't be deleted (probably because solution / template participations don't get deleted anymore?)
- [ ] When opening a text exercise, the following call fails: https://artemistest.ase.in.tum.de/api/text-editor/1530 => failed to lazily initialize a collection of role: de.tum.in.www1.artemis.domain.Participation.results, could not initialize proxy - no Session
- [ ] Migration scripts for setting discriminator needs to be run again on the test server, otherwise existing participations can't be retrieved.
